### PR TITLE
gadget: implement write content from gadget information

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
WriteContent will write gadget content to the relevant devices passed as gadget volumes. This will be called from the installation handlers.

### CODE REVIEW

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:439]: wouldn't /sys/class/block and the no [6:] work the same? without the counting?

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:414]: nitpick: I wouldn't contract Struct in function names, for args it's ok

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:477]: is this a comment or a TODO?

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:506]: should we validate somehow .Device before processing it?

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:471]: we need a TODO that this code will need attention for the encrypted partitons? as the device  there will not be a partition

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:462]: applyLayoutTo instead of build or something like that?  same not about Struct

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:465]: can't we read  partPath/partition instead ? or pass around the Disk we obtain onDiskVolumeFromPartitionSysfsPath and match through its Partitions?

**Review by pedronis — COMMENTED** [review]
did a pass, some comments/questions. Not entirely a fan of partIndexFromPartName

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:501]: yes, we will need to think about that

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:502]: this is not the check for raw vs not though, the check for that is whether the structure has a filesystem or not

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: I've added a TODO but right before calling `sysfsPathForBlockDevice()`, as the first access to device is there.

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Comment, I've expanded it a bit.

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Changed

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:485]: Somewhat we already validate it, first when we call Readlink (it will fail if not a block device), then when we try to find the disk, etc. Not sure if adding a test previous to that would make much sense.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks for the quick review, I've pushed now addressing your comments.

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:461]: do we need a unit test for this?

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:416]: related to the validation comment from before, seems then this error should be wrapped in something like "cannot find..." or "cannot resolve..." ?

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:485]: seems like this Base could/should go inside the helper itself? anyway Base is idempotent

File: gadget/install/install.go
Comment (pedronis) [inline gadget/install/install.go:443]: same for this one

**Review by pedronis — APPROVED** [review]
thanks, some smaller comments

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:461]: I've started with this but I think it will be better to have as a follow up if that's ok, as it involves quite a bit of mocking because of disk accesses.

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Done

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Done

File: gadget/install/content.go
Comment (mvo5) [inline gadget/install/content.go:81]: Thanks for adding this!

File: gadget/install/install.go
Comment (mvo5) [inline gadget/install/install.go:383]: The comment can now be removed, it only applied to the old code which was doing `filepath.Base(p.Node) == partDevice` here

File: gadget/install/install.go
Comment (mvo5) [inline gadget/install/install.go:395]: (nitpick) I prefer  long lines ;)

File: gadget/install/install.go
Comment (mvo5) [inline gadget/install/install.go:419]: Is this comment correct? When I look at my system I have:
```
$ ls -l /sys/block/sda
lrwxrwxrwx 1 root root 0 Sep 29 08:56 /sys/block/sda -> ../devices/pci0000:00/0000:00:01.3/0000:02:00.1/ata6/host5/target5:0:0/5:0:0:0/block/sda
```
i.e. only a single `../`.

File: gadget/install/install.go
Comment (mvo5) [inline gadget/install/install.go:420]: Should we also call filepath.Clean() here? I guess it does not really matter but filepath.EvalSymlinks() also does that (but we can't use that AIUI because it's a bit silly and works relative to CWD).

File: gadget/install/install.go
Comment (mvo5) [inline gadget/install/install.go:473]: maybe slightly more context what exactly there is to fix here :) ?

**Review by mvo5 — APPROVED** [review]
Thanks, this looks nice and very clean. Some inline suggestions/questions but nothing is a blocker.

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:419]: Right, but that is `/sys/block`, not `/sys/class/block`.

Comment (pedronis) [reply gadget/install/install.go:419]: this code really expects partitions, that will not be true for when we deal with encrpyted partitions otoh we have not content to write to those usually. I think this is ok to unblock us but we might need to rethink this a bit later

Comment (pedronis) [reply gadget/install/install.go:419]: @mvo is /sys/block/devices another symlink?

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Good catch, thanks!

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:417]: FTR `/sys/block` does not show partitions, only disks, while `/sys/class/block` does (there is no `/sys/block/devices`)

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:418]: It should not be necessary because of the way we obtain the path - if it actually was it would be probably a bug

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go:460]: Added

File: gadget/install/install.go
Comment (alfonsosanchezbeato) [reply gadget/install/install.go]: Indeed ;)

**Review by alfonsosanchezbeato — COMMENTED** [review]
Thanks for the reviews! I've addressed&answered the latest comments and merged master to get the fix for the static checks.

File: gadget/install/install.go
Comment (Meulengracht) [inline gadget/install/install.go:469]: I know you already have two approvals here, but I'd still want to reconsider this double nested for-loop. It makes the function huge and also makes it hard to read. Could you maybe split the contents of the `for _, volStruct := range vol.Structure` for loop out into a function, and then you wouldn't need to pre-declare the `onDiskVol` variable and add it half-way through the for-loop to the list.

Comment (pedronis) [reply gadget/install/install.go:469]: this is fine to consider for a follow up, we also need to add unit tests anyway

Comment (alfonsosanchezbeato) [reply gadget/install/install.go:469]: Yes, please leave that for a follow-up if @Meulengracht is fine with that. Note though that I'm not sure how slitting the inner loop would help with avoiding the pre-declartion of `onDiskVol` (you would need to do that in the helper function anyway).

**Pull request merged**
